tools/nl/embeddings/utils.py

```diff
@@ -13,146 +13,179 @@
-  # Determine the columns which were common.
-  common_cols = set()
-  for col in df_1.columns:
-    if col.endswith("_x") or col.endswith("_y"):
-      common_cols.add(col.replace("_x", "").replace("_y", ""))
-
-  # Replace the common columns with their concatenation.
-  for col in common_cols:
-    df_1[col] = df_1[f"{col}_x"].str.cat(df_1[f"{col}_y"], sep=";")
-    df_1[col] = df_1[col].replace(to_replace="^;", value="", regex=True)
-    df_1 = df_1.drop(columns=[f"{col}_x", f"{col}_y"])
-
-  return df_1
-
-
-def concat_alternatives(alternatives: List[str],
-                        max_alternatives,
-                        delimiter=";") -> str:
-  alts = set(alternatives[0:max_alternatives])
-  return f"{delimiter}".join(sorted(alts))
-
-
-def split_alt_string(alt_string: str) -> List[str]:
-  alts = []
-  for alt in alt_string.split(";"):
-    if alt:
-      alts.append(alt.strip())
-  return alts
-
-
-def add_sv(name: str, sv: str, text2sv: Dict[str, str],
-           dup_svs: List[List[str]]) -> None:
-  osv = text2sv.get(name)
-  if not osv or osv == sv:
-    text2sv[name] = sv
-    return
-
-  # This is a case of duplicate SV.  Prefer the non sdg, human-curated, shorter SV.
-  # Track it.
-  pref, drop = sv, osv
-  if sv.startswith('dc/topic/sdg'):
-    # sv is an sdg topic. Prefer osv if osv is not an sdg topic. Otherwise, go
-    # by dcid len.
-    if not osv.startswith('dc/topic/sdg') or len(osv) <= len(sv):
-      pref, drop = osv, sv
-  elif ((osv.startswith('dc/') and sv.startswith('dc/')) or
-        (not osv.startswith('dc/') and not sv.startswith('dc/'))):
-    # Both SVs are autogen or both aren't. Go by dcid len.
-    if len(osv) <= len(sv):
-      pref, drop = osv, sv
-  elif sv.startswith('dc/'):
-    # sv is autogen, prefer osv.
-    pref, drop = osv, sv
-
-  text2sv[name] = pref
-  dup_svs.append([pref, drop, name])
-
-
-def dedup_texts(df: pd.DataFrame) -> Tuple[Dict[str, str], List[List[str]]]:
-  """Dedup multiple texts mapped to the same DCID and return a list."""
-  text2sv_dict = {}
-  dup_sv_rows = [['PreferredSV', 'DroppedSV', 'DuplicateName']]
-  for _, row in df.iterrows():
-    sv = row[DCID_COL].strip()
-
-    # All alternative sentences are retrieved from COL_ALTERNATIVES, which
-    # are expected to be delimited by ";" (semi-colon).
-    if COL_ALTERNATIVES in row:
-      alternatives = row[COL_ALTERNATIVES].split(';')
-      alternatives = [a.strip() for a in alternatives if a.strip()]
-      for alt in alternatives:
-        add_sv(alt, sv, text2sv_dict, dup_sv_rows)
-
-  return (text2sv_dict, dup_sv_rows)
-
-
-def build_embeddings(
-    text2sv: Dict[str, str],
-    model: SentenceTransformer = None,
-    model_endpoint: aiplatform.Endpoint = None) -> pd.DataFrame:
-  """Builds the embeddings dataframe.
-
-  The output dataframe contains the embeddings columns (typically 384) + dcid + sentence.
+def build_preindex(fm: FileManager,
+                   save: bool = False) -> Tuple[List[str], List[str]]:
```

Comment:
nit:  given all of this will fit in memory, wonder if instead of `save` option to build_preindex, just have a write_preindex function?

Reply:
Hmm, actually who sets save to False?